deltacat/utils/daft.py

```diff
@@ -78,10 +78,35 @@ def daft_s3_file_to_table(
     logger.debug(f"Time to read S3 object from {s3_url} into daft table: {latency}s")
 
     if kwargs.get("schema") is not None:
-        schema = kwargs["schema"]
+        input_schema = kwargs["schema"]
         if include_columns is not None:
-            schema = pa.schema([schema.field(col) for col in include_columns])
-        daft_schema = Schema.from_pyarrow_schema(schema)
-        return table.cast_to_schema(daft_schema).to_arrow()
+            input_schema = pa.schema(
+                [input_schema.field(col) for col in include_columns]
+            )
+        elif column_names is not None:
+            input_schema = pa.schema([input_schema.field(col) for col in column_names])
+        input_schema_names = set(input_schema.names)
+
+        # Perform casting of types to provided schema's types
+        cast_to_schema = [
+            input_schema.field(inferred_field.name)
+            if inferred_field.name in input_schema_names
+            else inferred_field
+            for inferred_field in pa_table.schema
+        ]
+        casted_table = pa_table.cast(pa.schema(cast_to_schema))
+
+        # Reorder and pad columns with a null column where necessary
+        pa_table_column_names = set(casted_table.column_names)
+        columns = []
+        for name in input_schema.names:
+            if name in pa_table_column_names:
+                columns.append(casted_table[name])
+            else:
+                columns.append(
+                    pa.nulls(len(casted_table), type=input_schema.field(name).type)
+                )
+        schema = input_schema.with_metadata(pa_table.schema.metadata)
+        return pa.table(columns, schema=schema)
```

Comment:
This looks most accurate at this time. Would it be worthwhile to move this to separate function into pyarrow related utility?

Reply:
Yes, I can. Would you like to share this logic with the pyarrow reads as well?

Reply:
Yes, absolutely, but I can take that as part of another PR.